File: src/vibetotext/api_server.py
"""HTTP API server — exposes TTS and status endpoints for external tools."""

import json
import threading
from http.server import HTTPServer, BaseHTTPRequestHandler

from vibetotext.tts import speak

DEFAULT_PORT = 7865


class _Handler(BaseHTTPRequestHandler):
    """Handle API requests."""

    def log_message(self, format, *args):
        # Suppress default stderr logging
        pass

    def _send_json(self, status: int, data: dict):
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Access-Control-Allow-Origin", "*")
        self.end_headers()
        self.wfile.write(json.dumps(data).encode())

    def do_OPTIONS(self):
        """Handle CORS preflight."""
        self.send_response(204)
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "POST, GET, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type")
        self.end_headers()

    def do_GET(self):
        if self.path == "/api/status":
            self._send_json(200, {"status": "ok", "service": "vibetotext"})
        else:
            self._send_json(404, {"error": "not found"})

    def do_POST(self):
        if self.path == "/api/speak":
            self._handle_speak()
        else:
            self._send_json(404, {"error": "not found"})

    def _handle_speak(self):
        try:
            length = int(self.headers.get("Content-Length", 0))
            body = self.rfile.read(length)
            data = json.loads(body)
            text = data.get("text", "").strip()

            if not text:
                self._send_json(400, {"error": "missing 'text' field"})
                return

            speak(text)
            self._send_json(200, {"status": "speaking", "text": text})

        except json.JSONDecodeError:
            self._send_json(400, {"error": "invalid JSON"})
        except Exception as e:
            self._send_json(500, {"error": str(e)})


class ApiServer:
    """Lightweight HTTP server for TTS and status endpoints."""

    def __init__(self, port: int = DEFAULT_PORT):
        self.port = port
        self._server: HTTPServer | None = None
        self._thread: threading.Thread | None = None

    def start(self):
        try:
            self._server = HTTPServer(("127.0.0.1", self.port), _Handler)
            self._thread = threading.Thread(target=self._server.serve_forever, daemon=True)
            self._thread.start()
            print(f"[API] Server running on http://127.0.0.1:{self.port}")
            print(f"[API]   POST /api/speak  {{\"text\": \"...\"}}")
            print(f"[API]   GET  /api/status")
        except OSError as e:
            print(f"[API] Failed to start server on port {self.port}: {e}")

    def stop(self):
        if self._server:
            self._server.shutdown()
            self._server = None

    @property
    def url(self) -> str:
        return f"http://127.0.0.1:{self.port}"

File: windows-native/src/VibeToText/Core/ApiServer.cs
using System.Net;
using System.Text;
using System.Text.Json;

namespace VibeToText.Core;

/// <summary>
/// Lightweight HTTP API server for external integrations.
/// Listens on http://127.0.0.1:7865/ and exposes:
///   GET  /api/status → {"status": "ok", "service": "vibetotext"}
///   POST /api/speak  → {"text": "..."} → calls TtsService.Speak()
/// </summary>
public class ApiServer : IDisposable
{
    private readonly HttpListener _listener;
    private readonly TtsService _ttsService;
    private readonly int _port;
    private CancellationTokenSource? _cts;

    public ApiServer(TtsService ttsService, int port = 7865)
    {
        _ttsService = ttsService;
        _port = port;
        _listener = new HttpListener();
        _listener.Prefixes.Add($"http://127.0.0.1:{_port}/");
    }

    public void Start()
    {
        try
        {
            _cts = new CancellationTokenSource();
            _listener.Start();
            Console.WriteLine($"[API] Server listening on http://127.0.0.1:{_port}/");
            Task.Run(() => AcceptLoop(_cts.Token));
        }
        catch (Exception ex)
        {
            Console.WriteLine($"[API] Failed to start server: {ex.Message}");
        }
    }

    public void Stop()
    {
        try
        {
            _cts?.Cancel();
            _listener.Stop();
            Console.WriteLine("[API] Server stopped.");
        }
        catch { }
    }

    public void Dispose()
    {
        Stop();
        _cts?.Dispose();
        ((IDisposable)_listener).Dispose();
    }

    private async Task AcceptLoop(CancellationToken ct)
    {
        while (!ct.IsCancellationRequested)
        {
            try
            {
                var ctx = await _listener.GetContextAsync().WaitAsync(ct);
                // Handle each request in its own task (don't await — fire and forget)
                _ = Task.Run(() => HandleRequest(ctx), ct);
            }
            catch (OperationCanceledException)
            {
                break;
            }
            catch (HttpListenerException) when (ct.IsCancellationRequested)
            {
                break;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[API] Accept error: {ex.Message}");
            }
        }
    }

    private async Task HandleRequest(HttpListenerContext ctx)
    {
        var req = ctx.Request;
        var resp = ctx.Response;

        // Add CORS headers for local dev tools
        resp.Headers.Add("Access-Control-Allow-Origin", "*");
        resp.Headers.Add("Access-Control-Allow-Methods", "GET, POST, OPTIONS");
        resp.Headers.Add("Access-Control-Allow-Headers", "Content-Type");

        try
        {
            // Handle CORS preflight
            if (req.HttpMethod == "OPTIONS")
            {
                resp.StatusCode = 204;
                resp.Close();
                return;
            }

            var path = req.Url?.AbsolutePath ?? "";

            switch (path)
            {
                case "/api/status" when req.HttpMethod == "GET":
                    await WriteJson(resp, 200, new { status = "ok", service = "vibetotext" });
                    break;

                case "/api/speak" when req.HttpMethod == "POST":
                    await HandleSpeak(req, resp);
                    break;

                default:
                    await WriteJson(resp, 404, new { error = "not_found", message = $"No route for {req.HttpMethod} {path}" });
                    break;
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine($"[API] Request error: {ex.Message}");
            try
            {
                await WriteJson(resp, 500, new { error = "internal_error", message = ex.Message });
            }
            catch { }
        }
    }

    private async Task HandleSpeak(HttpListenerRequest req, HttpListenerResponse resp)
    {
        string body;
        using (var reader = new System.IO.StreamReader(req.InputStream, req.ContentEncoding))
        {
            body = await reader.ReadToEndAsync();
        }

        if (string.IsNullOrWhiteSpace(body))
        {
            await WriteJson(resp, 400, new { error = "bad_request", message = "Empty request body" });
            return;
        }

        JsonDocument? doc = null;
        try
        {
            doc = JsonDocument.Parse(body);
        }
        catch (JsonException)
        {
            await WriteJson(resp, 400, new { error = "bad_request", message = "Invalid JSON" });
            return;
        }

        var text = doc.RootElement.TryGetProperty("text", out var textProp) ? textProp.GetString() : null;
        doc.Dispose();

        if (string.IsNullOrWhiteSpace(text))
        {
            await WriteJson(resp, 400, new { error = "bad_request", message = "Missing or empty 'text' field" });
            return;
        }

        _ttsService.Speak(text);
        await WriteJson(resp, 200, new { status = "speaking" });
    }

    private static async Task WriteJson(HttpListenerResponse resp, int statusCode, object data)
    {
        resp.StatusCode = statusCode;
        resp.ContentType = "application/json";
        var json = JsonSerializer.Serialize(data);
        var bytes = Encoding.UTF8.GetBytes(json);
        resp.ContentLength64 = bytes.Length;
        await resp.OutputStream.WriteAsync(bytes);
        resp.Close();
    }
}
